File: src/main/java/com/parkjaehwan/addressbench/Lcs.java
package com.parkjaehwan.addressbench;

/**
 * 최장 공통 부분열(LCS). 두 행만 쓰는 O(min(m,n)) 메모리 DP.
 */
public final class Lcs {
    private Lcs() {
    }

    public static int length(CharSequence left, CharSequence right) {
        if (left.length() > right.length()) {
            CharSequence swap = left;
            left = right;
            right = swap;
        }
        int m = left.length();
        int[] previous = new int[m + 1];
        int[] current = new int[m + 1];
        for (int j = 0; j < right.length(); j++) {
            char rc = right.charAt(j);
            current[0] = 0;
            for (int i = 1; i <= m; i++) {
                if (left.charAt(i - 1) == rc) {
                    current[i] = previous[i - 1] + 1;
                } else {
                    current[i] = Math.max(previous[i], current[i - 1]);
                }
            }
            int[] swap = previous;
            previous = current;
            current = swap;
        }
        return previous[m];
    }

    /** 정규화한 두 문자열의 LCS 길이를 긴 쪽 길이로 나눈 값. 둘 다 비어 있으면 1.0. */
    public static double similarity(String left, String right) {
        return similarityNormalized(Normalizer.normalize(left), Normalizer.normalize(right));
    }

    /** 이미 정규화된 두 문자열에 대한 유사도. 전수 탐색 루프에서 쓴다. */
    public static double similarityNormalized(CharSequence left, CharSequence right) {
        int denominator = Math.max(left.length(), right.length());
        return denominator == 0 ? 1.0 : (double) length(left, right) / denominator;
    }
}

File: src/address_benchmark/lcs.py
from dataclasses import dataclass
from heapq import heappush, heapreplace
from time import perf_counter_ns

from .normalize import normalize


def lcs_length(left: str, right: str) -> int:
    if len(left) > len(right):
        left, right = right, left
    previous = [0] * (len(left) + 1)
    for right_char in right:
        current = [0]
        for index, left_char in enumerate(left, 1):
            current.append(previous[index - 1] + 1 if left_char == right_char else max(previous[index], current[-1]))
        previous = current
    return previous[-1]


def similarity(left: str, right: str) -> float:
    left, right = normalize(left), normalize(right)
    denominator = max(len(left), len(right))
    return lcs_length(left, right) / denominator if denominator else 1.0


@dataclass(frozen=True)
class Match:
    address_id: int
    alias: str
    score: float


class ExhaustiveLcsMatcher:
    """Intentionally simple O(N * query_length * alias_length) baseline."""

    def __init__(self, records: list[tuple[int, str]]):
        self.records = records

    def search(self, query: str, top_k: int = 5) -> tuple[list[Match], int]:
        started = perf_counter_ns()
        heap: list[tuple[float, int, str]] = []
        for address_id, alias in self.records:
            item = (similarity(query, alias), address_id, alias)
            if len(heap) < top_k:
                heappush(heap, item)
            elif item > heap[0]:
                heapreplace(heap, item)
        matches = [Match(address_id, alias, score) for score, address_id, alias in sorted(heap, reverse=True)]
        return matches, perf_counter_ns() - started

